src/XMakeBuildEngine/Evaluation/MSBuildExtensionsPathHelper.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            var extnPaths = new List<string>();
+            var refKindAsString = ReferenceKindToString(refKind);
+            var envVar = Environment.GetEnvironmentVariable(refKindAsString + "SearchPaths");
+
+            var allPaths = String.Join(Path.PathSeparator.ToString(),
+                                        //FIXME: IsOSX returns false :/
+                                        NativeMethodsShared.IsUnix ? s_osxExternalXBuildDir : String.Empty,
+                                        envVar ?? String.Empty);
+
+            var pathsTable = new Dictionary<string, string>();
+            foreach(var extnPath in allPaths.Split(new char[]{Path.PathSeparator}, StringSplitOptions.RemoveEmptyEntries)) {
+                if(pathsTable.ContainsKey(extnPath))
+                    continue;
+
+                if(!Directory.Exists(extnPath))
+                    continue;
+
+                pathsTable[extnPath] = extnPath;
+                extnPaths.Add(extnPath);
+            }
+
+            return extnPaths;
+        }
+
+        public static MSBuildExtensionsPathReferenceKind GetReferenceKindIn(string expression)
+        {
+            if(expression.IndexOf("$(MSBuildExtensionsPath)") >= 0)
+                return MSBuildExtensionsPathReferenceKind.Default;
+
+            if(expression.IndexOf("$(MSBuildExtensionsPath32)") >= 0)
+                return MSBuildExtensionsPathReferenceKind.Path32;
+
+            if(expression.IndexOf("$(MSBuildExtensionsPath64)") >= 0)
+                return MSBuildExtensionsPathReferenceKind.Path64;
+
+            return MSBuildExtensionsPathReferenceKind.None;
+        }
+
+        public static string ReferenceKindToString(MSBuildExtensionsPathReferenceKind refKind)
+        {
+            switch(refKind) {
+                case MSBuildExtensionsPathReferenceKind.Default:
+                    return "MSBuildExtensionsPath";
+                case MSBuildExtensionsPathReferenceKind.Path32:
+                    return "MSBuildExtensionsPath32";
+                case MSBuildExtensionsPathReferenceKind.Path64:
+                    return "MSBuildExtensionsPath64";
+                case MSBuildExtensionsPathReferenceKind.None:
+                default:
+                    return null;
+            }
+        }
+
+        ///<summary>
+        /// Helper method for resetting the lists for testing, since we need to run the
+        /// tests with different values for the *SearchPaths
+        ///</summary>
+        private static void ResetPathsListsForTesting()
+        {
+            s_pathsLists = new Lazy<ExtensionsPathsLists>(true);
+        }
+
+        class ExtensionsPathsLists
+        {
+            /// <summary>
+            /// Cached list of extension paths to be used as search paths for $(MSBuildExtensionsPath)
+            /// </summary>
+            public readonly List<string> Default;
```

Comment:
You could convert these 3 fields into a single Dictionary<MSBuildExtensionsPathReferenceKind, List<string>> and have GetFallbackExtensionsPathsFor delegate here instead of having that switch statement